File: src/test/java/com/joecollins/graphics/components/SwingFrameTest.java
package com.joecollins.graphics.components;

import static com.joecollins.graphics.utils.RenderTestUtils.compareRendering;
import static org.junit.Assert.assertEquals;

import com.joecollins.bindings.Binding;
import java.awt.Color;
import java.io.IOException;
import org.junit.Test;

public class SwingFrameTest {

  @Test
  public void testSwingRange() {
    SwingFrame frame = new SwingFrame();
    frame.setRangeBinding(Binding.fixedBinding(10));
    assertEquals(10, frame.getRange());
  }

  @Test
  public void testSwingValue() {
    SwingFrame frame = new SwingFrame();
    frame.setValueBinding(Binding.fixedBinding(3));
    assertEquals(3, frame.getValue());
  }

  @Test
  public void testLeftRightColors() {
    SwingFrame frame = new SwingFrame();
    frame.setLeftColorBinding(Binding.fixedBinding(Color.BLUE));
    frame.setRightColorBinding(Binding.fixedBinding(Color.RED));
    assertEquals(Color.BLUE, frame.getLeftColor());
    assertEquals(Color.RED, frame.getRightColor());
  }

  @Test
  public void testBottomText() {
    SwingFrame frame = new SwingFrame();
    frame.setBottomTextBinding(Binding.fixedBinding("4.7% SWING LIB TO CON"));
    frame.setBottomColorBinding(Binding.fixedBinding(Color.BLUE));
    assertEquals("4.7% SWING LIB TO CON", frame.getBottomText());
    assertEquals(Color.BLUE, frame.getBottomColor());
  }

  @Test
  public void testRenderNoSwing() throws IOException {
    SwingFrame frame = new SwingFrame();
    frame.setHeaderBinding(Binding.fixedBinding("SWING SINCE 2015"));
    frame.setRangeBinding(Binding.fixedBinding(10));
    frame.setValueBinding(Binding.fixedBinding(0));
    frame.setLeftColorBinding(Binding.fixedBinding(Color.BLUE));
    frame.setRightColorBinding(Binding.fixedBinding(Color.RED));
    frame.setBottomTextBinding(Binding.fixedBinding("WAITING FOR RESULTS..."));
    frame.setBottomColorBinding(Binding.fixedBinding(Color.BLACK));
    frame.setSize(256, 128);

    compareRendering("SwingFrame", "NoSwing", frame);
  }

  @Test
  public void testRenderSwingRight() throws IOException {
    SwingFrame frame = new SwingFrame();
    frame.setHeaderBinding(Binding.fixedBinding("SWING SINCE 2015"));
    frame.setRangeBinding(Binding.fixedBinding(10));
    frame.setValueBinding(Binding.fixedBinding(4.7));
    frame.setLeftColorBinding(Binding.fixedBinding(Color.BLUE));
    frame.setRightColorBinding(Binding.fixedBinding(Color.RED));
    frame.setBottomTextBinding(Binding.fixedBinding("4.7% SWING LIB TO CON"));
    frame.setBottomColorBinding(Binding.fixedBinding(Color.BLUE));
    frame.setSize(256, 128);

    compareRendering("SwingFrame", "SwingRight", frame);
  }

  @Test
  public void testRenderSwingLeft() throws IOException {
    SwingFrame frame = new SwingFrame();
    frame.setHeaderBinding(Binding.fixedBinding("SWING SINCE 2015"));
    frame.setRangeBinding(Binding.fixedBinding(10));
    frame.setValueBinding(Binding.fixedBinding(-1.3));
    frame.setLeftColorBinding(Binding.fixedBinding(Color.RED));
    frame.setRightColorBinding(Binding.fixedBinding(Color.ORANGE));
    frame.setBottomTextBinding(Binding.fixedBinding("1.3% SWING LIB TO NDP"));
    frame.setBottomColorBinding(Binding.fixedBinding(Color.ORANGE));
    frame.setSize(256, 128);

    compareRendering("SwingFrame", "SwingLeft", frame);
  }

  @Test
  public void testRenderMaxSwingRight() throws IOException {
    SwingFrame frame = new SwingFrame();
    frame.setHeaderBinding(Binding.fixedBinding("SWING SINCE 2015"));
    frame.setRangeBinding(Binding.fixedBinding(10));
    frame.setValueBinding(Binding.fixedBinding(19.9));
    frame.setLeftColorBinding(Binding.fixedBinding(Color.BLUE));
    frame.setRightColorBinding(Binding.fixedBinding(Color.RED));
    frame.setBottomTextBinding(Binding.fixedBinding("19.9% SWING LIB TO CON"));
    frame.setBottomColorBinding(Binding.fixedBinding(Color.BLUE));
    frame.setSize(256, 128);

    compareRendering("SwingFrame", "MaxSwingRight", frame);
  }

  @Test
  public void testRenderMaxSwingLeft() throws IOException {
    SwingFrame frame = new SwingFrame();
    frame.setHeaderBinding(Binding.fixedBinding("SWING SINCE 2015"));
    frame.setRangeBinding(Binding.fixedBinding(10));
    frame.setValueBinding(Binding.fixedBinding(-21.6));
    frame.setLeftColorBinding(Binding.fixedBinding(Color.RED));
    frame.setRightColorBinding(Binding.fixedBinding(Color.GREEN));
    frame.setBottomTextBinding(Binding.fixedBinding("21.6% SWING LIB TO GRN"));
    frame.setBottomColorBinding(Binding.fixedBinding(Color.GREEN));
    frame.setSize(256, 128);

    compareRendering("SwingFrame", "MaxSwingLeft", frame);
  }

  @Test
  public void testRenderAccents() throws IOException {
    SwingFrame frame = new SwingFrame();
    frame.setHeaderBinding(Binding.fixedBinding("CHANGES APR\u00c8S 2014"));
    frame.setRangeBinding(Binding.fixedBinding(10));
    frame.setValueBinding(Binding.fixedBinding(0));
    frame.setLeftColorBinding(Binding.fixedBinding(Color.BLUE));
    frame.setRightColorBinding(Binding.fixedBinding(Color.RED));
    frame.setBottomTextBinding(Binding.fixedBinding("VOIX PAS R\u00c9\u00c7US"));
    frame.setBottomColorBinding(Binding.fixedBinding(Color.BLACK));
    frame.setSize(256, 128);

    compareRendering("SwingFrame", "Accents", frame);
  }
}

File: src/test/java/com/joecollins/graphics/components/SwingFrameTest.kt
package com.joecollins.graphics.components

import com.joecollins.graphics.utils.RenderTestUtils.compareRendering
import com.joecollins.pubsub.asOneTimePublisher
import org.awaitility.Awaitility
import org.hamcrest.core.IsEqual
import org.junit.Assert
import org.junit.Test
import java.awt.Color
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.Throws

class SwingFrameTest {
    @Test
    fun testSwingRange() {
        val frame = SwingFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 3.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "4.7% SWING LIB TO CON".asOneTimePublisher(),
            bottomColorPublisher = Color.BLUE.asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getRange() }, IsEqual(10))
    }

    @Test
    fun testSwingValue() {
        val frame = SwingFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 3.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "4.7% SWING LIB TO CON".asOneTimePublisher(),
            bottomColorPublisher = Color.BLUE.asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getValue() }, IsEqual(3))
    }

    @Test
    fun testLeftRightColors() {
        val frame = SwingFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 3.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "4.7% SWING LIB TO CON".asOneTimePublisher(),
            bottomColorPublisher = Color.BLUE.asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getLeftColor() }, IsEqual(Color.BLUE))
        Assert.assertEquals(Color.RED, frame.getRightColor())
    }

    @Test
    fun testBottomText() {
        val frame = SwingFrame(
            headerPublisher = (null as String?).asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 3.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "4.7% SWING LIB TO CON".asOneTimePublisher(),
            bottomColorPublisher = Color.BLUE.asOneTimePublisher()
        )
        Awaitility.await().atMost(500, TimeUnit.MILLISECONDS)
            .until({ frame.getBottomColor() }, IsEqual(Color.BLUE))
        Assert.assertEquals("4.7% SWING LIB TO CON", frame.getBottomText())
    }

    @Test
    @Throws(IOException::class)
    fun testRenderNoSwing() {
        val frame = SwingFrame(
            headerPublisher = "SWING SINCE 2015".asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 0.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "WAITING FOR RESULTS...".asOneTimePublisher(),
            bottomColorPublisher = Color.BLACK.asOneTimePublisher()
        )
        frame.setSize(256, 128)
        compareRendering("SwingFrame", "NoSwing", frame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderSwingRight() {
        val frame = SwingFrame(
            headerPublisher = "SWING SINCE 2015".asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 4.7.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "4.7% SWING LIB TO CON".asOneTimePublisher(),
            bottomColorPublisher = Color.BLUE.asOneTimePublisher()
        )
        frame.setSize(256, 128)
        compareRendering("SwingFrame", "SwingRight", frame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderSwingLeft() {
        val frame = SwingFrame(
            headerPublisher = "SWING SINCE 2015".asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = (-1.3).asOneTimePublisher(),
            leftColorPublisher = Color.RED.asOneTimePublisher(),
            rightColorPublisher = Color.ORANGE.asOneTimePublisher(),
            bottomTextPublisher = "1.3% SWING LIB TO NDP".asOneTimePublisher(),
            bottomColorPublisher = Color.ORANGE.asOneTimePublisher()
        )
        frame.setSize(256, 128)
        compareRendering("SwingFrame", "SwingLeft", frame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderMaxSwingRight() {
        val frame = SwingFrame(
            headerPublisher = "SWING SINCE 2015".asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 19.9.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "19.9% SWING LIB TO CON".asOneTimePublisher(),
            bottomColorPublisher = Color.BLUE.asOneTimePublisher()
        )
        frame.setSize(256, 128)
        compareRendering("SwingFrame", "MaxSwingRight", frame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderMaxSwingLeft() {
        val frame = SwingFrame(
            headerPublisher = "SWING SINCE 2015".asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = (-21.6).asOneTimePublisher(),
            leftColorPublisher = Color.RED.asOneTimePublisher(),
            rightColorPublisher = Color.GREEN.asOneTimePublisher(),
            bottomTextPublisher = "21.6% SWING LIB TO GRN".asOneTimePublisher(),
            bottomColorPublisher = Color.GREEN.asOneTimePublisher()
        )
        frame.setSize(256, 128)
        compareRendering("SwingFrame", "MaxSwingLeft", frame)
    }

    @Test
    @Throws(IOException::class)
    fun testRenderAccents() {
        val frame = SwingFrame(
            headerPublisher = "CHANGES APR\u00c8S 2014".asOneTimePublisher(),
            rangePublisher = 10.asOneTimePublisher(),
            valuePublisher = 0.asOneTimePublisher(),
            leftColorPublisher = Color.BLUE.asOneTimePublisher(),
            rightColorPublisher = Color.RED.asOneTimePublisher(),
            bottomTextPublisher = "VOIX PAS R\u00c9\u00c7US".asOneTimePublisher(),
            bottomColorPublisher = Color.BLACK.asOneTimePublisher()
        )
        frame.setSize(256, 128)
        compareRendering("SwingFrame", "Accents", frame)
    }
}
